src/napari_time_slicer/__init__.py

```diff
@@ -51,6 +56,76 @@ def worker_function(*args, **kwargs):
 
     return worker_function
 
+@curry
+def time_slicer_dask(function: Callable) -> Callable:
+    @wraps(function)
+    def worker_function(*args, **kwargs):
+        args = list(args)
+        sig = inspect.signature(function)
+        # create mapping from position and keyword arguments to parameters
+        # will raise a TypeError if the provided arguments do not match the signature
+        # https://docs.python.org/3/library/inspect.html#inspect.Signature.bind
+        bound = sig.bind(*args, **kwargs)
+        # set default values for missing arguments
+        # https://docs.python.org/3/library/inspect.html#inspect.BoundArguments.apply_defaults
+        bound.apply_defaults()
+
+        # Retrieve the viewer parameter so that we can know which current timepoint is selected
+        viewer = None
+        for key, value in bound.arguments.items():
+            if isinstance(value, napari.Viewer):
+                viewer = value
+
+        if viewer is None:
+            pass
+        else:
+            # in case of 4D-data (timelapse) lazily process each 3D frame and add to
+            # dask stack
+            if viewer.dims.ndim == 4:
+
+                # get sample of the output
+                _break_down_4d_to_2d_kwargs(bound.arguments, 0, viewer)
+                output_sample = function(*bound.args, **bound.kwargs)
+
+                # set up lazy processing
+                lazy_processed_image = delayed(function)
+
+                # make the dask stack
+                lazy_arrays = []
+                for i in range(viewer.dims.nsteps[0]):
+                    # needs to be called everytime we want to change the bound args
+                    bound = sig.bind(*args, **kwargs)
+                    bound.apply_defaults()
+                    _break_down_4d_to_2d_kwargs(bound.arguments, i, viewer)
+
+                    lazy_arrays.append(
+                        lazy_processed_image(*bound.args, **bound.kwargs)
+                    )
+
+                dask_arrays = [
+                    [da.from_delayed(
+                        delayed_reader, 
+                        shape=output_sample.shape, 
+                        dtype=output_sample.dtype)] 
+                    if len(output_sample.shape) == 2
+                    else da.from_delayed(
+                        delayed_reader, 
+                        shape=output_sample.shape, 
+                        dtype=output_sample.dtype
+                    )
+                    for delayed_reader in lazy_arrays
+                ]
+                # Stack into one large dask.array
+                stack = da.stack(
+                    dask_arrays, 
+                    axis=0)
+                return stack
+
+        # call the decorated function
+        result = function(*bound.args, **bound.kwargs)
```

Comment:
Hi Ryan @Cryaaa ,

that's an amazing PR! Before I try it out, one quick question: I'm confused about this line. Are you sure it's necessary?

Big thanks!
Robert

Reply:
Hey @haesleinhuepf 
I think this line is necessary if you still want to be able to apply the function you have wrapped to non-timelapse data. This functionality exists with your previous implementation and I did not want to break this functionality.